okhttp/build.gradle

```diff
@@ -1,9 +1,22 @@
 apply plugin: 'me.champeau.gradle.japicmp'
 
 jar {
-  manifest {
-    attributes('Automatic-Module-Name': 'okhttp3')
-  }
+  // modify these lines for MANIFEST.MF properties or for specific bnd instructions
+  // internal package needs to be exported too because interceptors use internal packages
+  bnd '''
+  Export-Package: \
+  okhttp3,\
+  okhttp3.internal.*
```

Comment:
Is there a concept of a friend package?

Reply:
See okhttp-tls comment

Reply:
in zipkin we do things like this..

export your internal package..
https://github.com/openzipkin/brave/blob/master/brave/bnd.bnd#L9

explicitly import when internal is used
https://github.com/openzipkin/brave/blob/master/context/slf4j/bnd.bnd#L1

block import when something is source retention..
https://github.com/openzipkin/brave/blob/master/instrumentation/http/bnd.bnd#L2

Reply:
Just to be sure I understand the approach you set the "braveinternal" attribute mandatory which enforces that the "braveinternal" attribute is set for every Import-Package statement? Therefore if someone wants to Import one of the internal packages the "braveinternal" attribute must explicitly set otherwise the bundle would not resolve?

So for my understanding this would be like a "I agree that what I'm importing is not public API and can change at will". I kind of like that approach because it prevents accidently importing a package which you shouldn't because it's internal.

Reply:
I think you understand it, yes. We have had bugs where we forgot to explicitly import the internal thing, but I think it is a better optimization anyway.

Reply:
I did hope that bnd does not specify mandatory attributes automatically but I have just tried it and bnd does in fact automatically infer mandatory attributes for Import-Package statements. Therefore if someone is using bnd as a build system he wouldn't even notice that there is an mandatory attribute called "okhttpinternal"/"braveinternal"...

Reply:
trying to understand the impact of this.. third parties outside this repo shouldn't be using the internal code anyway, so should be not impacted by ignorance of them. Only here, we have to make sure the bnds are coherent.. am I missing something?

Reply:
We want to ensure that only the okhttp sibling libraries are able to use the internal package of the main okhttp project. Therefore we add a mandatory "okhttpinternal" directive which ensures that for each Import-Package: okhttp3.internal statement the attribute must be explicitly defined. For us it's easy to define that attribute for the siblings okhttp3 libraries which use those internal packages. If a 3rdparty consumer would also import one of the internal packages without defining the attribute the bundle would resolve with an error according to OSGi spec because the attribute is not defined.
Unfortunately if the 3rdparty is also using bnd as a build system (like we in our internal project for example) we could freely use the okhttp3.internal package (even though we shouldn't) and we wouldn't event notice that we should define a mandatory attribute because bnd figures that out automatically.
As an example: I tried to add that attribute to the root okhttp3 project and I didn't have to specify the attribute in the sibling okhttp-sse library manually because bnd automatically inferred the mandatory attribute which kind of destroys the purpose of those whole attribute.
<pre>
Manifest-Version: 1.0
Automatic-Module-Name: okhttp3.sse
Bnd-LastModified: 1601454505191
Bundle-ManifestVersion: 2
Bundle-Name: com.squareup.okhttp3.sse
Bundle-SymbolicName: com.squareup.okhttp3.sse
Bundle-Version: 4.10.0.SNAPSHOT
Created-By: 14.0.2 (AdoptOpenJDK)
Export-Package: okhttp3.sse;uses:="kotlin,kotlin.jvm,okhttp3";version=
 "4.10.0"
Import-Package: kotlin;version="[1.4,2)",kotlin.io;version="[1.4,2)",k
 otlin.jvm;version="[1.4,2)",kotlin.jvm.internal;version="[1.4,2)",okh
 ttp3;version="[4.10,5)",okhttp3.internal;version="[4.10,5)";okhttpint
 ernal=true,okhttp3.internal.connection;version="[4.10,5)";okhttpinter
 nal=true,okio
Private-Package: okhttp3.internal.sse
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-5.1.2.202007211702
</pre>

Reply:
If this automatically sucks in the internal stuff, it would seem to be a behavioral difference in manifest generation here vs for example maven-bundle-plugin. This might be a different interpretation of what is sane behavior, or simply an accident in one or the other.

In any case, would you agree that the internal attribution is important to annotate even the gradle plugin or default configuration thwarts that guard rail?

cc also @helgoboss for input as it is sometimes hard for project teams to come to a sane approach for things like this if they aren't actively using the tooling.

Reply:
I have added the attribute although when bnd is used the benefit is negligible. I did not have to explicitly define the import in the sibling libraries because bnd adds the imports with the property set to true automatically...